File: java/tsp.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

class Point {
    public double x;
    public double y;

    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public String toString() {
        return String.format("Point (%f, %f)", x, y);
    }
}

class Tsp {

    private static double pathLength(Point start, Point[] path) {
        var length = 0.0;
        length += euclideanDistance(start, path[0]);
        for (int i=1; i<path.length; i++)
        {
            length += euclideanDistance(path[i - 1], path[i]);
        }
        length += euclideanDistance(path[path.length - 1], start);
        return length;
    }

    private static double euclideanDistance(Point a, Point b) {
        return Math.sqrt((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y));
    }

    private static Point[] getPoints(String filename) throws IOException {
        return Files.lines(Paths.get(filename)).map(x -> {
            var s = x.split(",");
            var p = new Point(Double.parseDouble(s[0]), Double.parseDouble(s[1]));
            return p;
        }).toArray(Point[]::new);
    }

    private static void bruteForceTSP(String filename) throws IOException {
        var points = getPoints(filename);
        var startPoint = points[0];

        var start = System.currentTimeMillis();
        var n = points.length;

        var indexes = new int[n];

        var shortestPath = points;
        var shortestLength = pathLength(startPoint, points);

        var i = 0;
        while (i < n)
        {
            if (indexes[i] < i)
            {
                var toSwapIndex = 0;
                if (!(i % 2 == 0)) {
                    toSwapIndex = indexes[i];
                }
                var tmp = points[toSwapIndex];
                points[toSwapIndex] = points[i];
                points[i] = tmp;

                var length = pathLength(startPoint, points);
                if (length < shortestLength)
                {
                    shortestLength = length;
                    shortestPath = points;
                }
                indexes[i]++;
                i = 0;
            }
            else
            {
                indexes[i] = 0;
                i++;
            }
        }
        var execTime = System.currentTimeMillis() - start;
        System.out.printf("{\"lang\" : \"java\", \"time\" : %d, \"distance\" :  %.14f}\n", execTime, shortestLength);
    }

    public static void main(String[] args) throws IOException {
        if (args.length != 1) {
            System.out.println("Please provide filename as argument");
            return;
        }
        
        bruteForceTSP(args[0]);
    }
}

File: kotlin/tsp.kt
import kotlin.collections.*
import kotlin.system.*
import java.io.File

data class Point(val x: Double, val y: Double)

fun getPoints(filename: String): ArrayList<Point> {
    var ret = ArrayList<Point>()
    File(filename).forEachLine {
        val s = it.split(",")
        val p = Point(s[0].toDouble(), s[1].toDouble())
        ret.add(p)
    }
    return ret
}

fun euclideanDistance(a: Point, b: Point): Double {
	return Math.sqrt((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y))
}

fun pathLength(start: Point, path: Array<Point>): Double {
	var length = 0.0;
	length += euclideanDistance(start, path.get(0));
	for (i in 1..path.size-1)
	{
		length += euclideanDistance(path.get(i - 1), path.get(i));
	}
	length += euclideanDistance(path.get(path.size - 1), start);
	return length;
}

fun bruteForceTSP(filename: String)
{
	var points: Array<Point> = getPoints(filename).toTypedArray()
    var startPoint = points[0]

    val start = System.currentTimeMillis()
    val n = points.size

    var indexes = Array<Int>(n) { 0 }

    var shortestPath: Array<Point> = points
    var shortestLength = pathLength(startPoint, points)

    var i = 0
    while (i < n)
    {
        if (indexes[i] < i)
        {
            val toSwapIndex = if (i % 2 == 0) 0 else indexes[i]
            val tmp = points[toSwapIndex]
            points[toSwapIndex] = points[i]
            points[i] = tmp

            val length = pathLength(startPoint, points)
            if (length < shortestLength)
            {
                shortestLength = length
                shortestPath = points
            }
            indexes[i]++
            i = 0
        }
        else
        {
            indexes[i] = 0
            ++i
        }
    }
    val execTime = System.currentTimeMillis() - start
    println("{\"lang\" : \"kotlin\", \"time\" : $execTime, \"distance\" :  $shortestLength}")
}

fun main(args: Array<String>)
{
    if (args.size < 1) {
        println("Please provide filename as argument")
        return
    }

	bruteForceTSP(args[0]);
}
